Preserve insertion order for extra dict items in assertion output

RafiB982 opened the pull request:
When dict assertions fail, pytest currently displays extra items using
unordered set iteration, which loses insertion order.

This change preserves insertion order for the
"Left contains N more items" and "Right contains N more items" sections
by iterating keys in mapping order instead of set order.

A focused regression test is included.

Closes #13503
<!--
Thanks for submitting a PR, your contribution is really appreciated!

Here is a quick checklist that should be present in PRs.

- [ ] Include documentation when adding new features.
- [X ] Include new tests or update existing tests when applicable.
- [X] Allow maintainers to push and squash when merging my commits. Please uncheck this if you prefer to squash the commits yourself.

If this change fixes an issue, please:

- [X ] Add text like ``closes #XYZW`` to the PR description and/or commits (where ``XYZW`` is the issue number). See the [github docs](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword) for more information.

Unless your change is trivial or a small documentation fix (e.g., a typo or reword of a small section) please:

- [ X] Create a new changelog file in the `changelog` folder, with a name like `<ISSUE NUMBER>.<TYPE>.rst`. See [changelog/README.rst](https://github.com/pytest-dev/pytest/blob/main/changelog/README.rst) for details.

  Write sentences in the **past or present tense**, examples:

  * *Improved verbose diff output with sequences.*
  * *Terminal summary statistics now use multiple colors.*

  Also make sure to end the sentence with a `.`.

- [ X] Add yourself to `AUTHORS` in alphabetical order.
-->

nicoddemus commented on src/_pytest/assertion/util.py:
```diff
@@ -520,7 +520,7 @@ def _compare_eq_dict(
                 + highlighter(saferepr({k: right[k]}))
             ]
-    extra_left = set_left - set_right
+    extra_left = [k for k in left if k not in right]
```
It is more efficient to reuse the set for the containment check:


```suggestion
    extra_left = [k for k in left if k not in set_right]
```

nicoddemus commented on testing/test_assertion.py:
```diff
@@ -2198,3 +2198,38 @@ def test_vvv():
+        test_order="""
+        def test_order():
+            a = {
+                "first": 1,
```
Please let's use a dict with more keys for this test, only 2 seems few to catch any regressions. I suggest using simply `"a"`, `"b"`, etc, inserted in non-alphabetical order.

nicoddemus reviewed (changes requested):
Thanks @RafiB982, please take a look at my comments.

RafiB982 commented:
Thanks for the review @nicoddemus 

I’ve addressed the feedback by:

1) reusing sets for containment checks in the implementation

2) expanding the test to use more keys in non-alphabetical insertion order

3) relying on -vv full assertion diffs and result.stdout.fnmatch_lines for verification

4) The rebase conflict has been resolved and tox -e py310 passes successfully.

Please let me know if you’d like any further adjustments.

nicoddemus commented on changelog/13503.bugfix.rst:
```diff
@@ -0,0 +1 @@
+Preserved insertion order for extra dictionary items in assertion failure output.
```
```suggestion
Display dictionary differences in assertion failures using the original key insertion order instead of sorted order.
```

nicoddemus commented:
The documentation failure is being addressed in https://github.com/pytest-dev/pytest/pull/14067.

RafiB982 commented:
Hi @nicoddemus ,
I’ve refreshed and rebased the branch onto the latest main and pushed the updated history.

RafiB982 commented:
Hello @nicoddemus ,
i am confused to see those many files under the PR i raised, as i did not make any changes.. 

(pytest-venv) rafi@ENDEAVOUR:~/pytest$ git show
commit 2ac996017eaa065271e6451757f7cf4f4c542950 (HEAD -> preserve-dict-insertion-order)
Author: Rafi <rafi_bigdata@outlook.com>
Date:   Fri Dec 26 18:14:14 2025 +0530

    Preserve insertion order for extra dict items in assertion output

**diff --git a/src/_pytest/assertion/util.py b/src/_pytest/assertion/util.py**
index f35d83a6f..a5bcc5d02 100644
--- a/src/_pytest/assertion/util.py
+++ b/src/_pytest/assertion/util.py
@@ -520,7 +520,7 @@ def _compare_eq_dict(
                 + " != "
                 + highlighter(saferepr({k: right[k]}))
             ]
-    extra_left = set_left - set_right
+    extra_left = [k for k in left if k not in right]
     len_extra_left = len(extra_left)
     if len_extra_left:
         explanation.append(
@@ -529,7 +529,7 @@ def _compare_eq_dict(
         explanation.extend(
             highlighter(pprint.pformat({k: left[k] for k in extra_left})).splitlines()
         )
-    extra_right = set_right - set_left
+    extra_right = [k for k in right if k not in left]
     len_extra_right = len(extra_right)
     if len_extra_right:
         explanation.append(
**diff --git a/testing/test_assertion.py b/testing/test_assertion.py**
:...skipping...
commit 2ac996017eaa065271e6451757f7cf4f4c542950 (HEAD -> preserve-dict-insertion-order)
Author: Rafi <rafi_bigdata@outlook.com>
Date:   Fri Dec 26 18:14:14 2025 +0530

    Preserve insertion order for extra dict items in assertion output

**diff --git a/src/_pytest/assertion/util.py b/src/_pytest/assertion/util.py**
index f35d83a6f..a5bcc5d02 100644
--- a/src/_pytest/assertion/util.py
+++ b/src/_pytest/assertion/util.py
@@ -520,7 +520,7 @@ def _compare_eq_dict(
                 + " != "
                 + highlighter(saferepr({k: right[k]}))
             ]
-    extra_left = set_left - set_right
+    extra_left = [k for k in left if k not in right]
     len_extra_left = len(extra_left)
     if len_extra_left:
         explanation.append(
@@ -529,7 +529,7 @@ def _compare_eq_dict(
         explanation.extend(
             highlighter(pprint.pformat({k: left[k] for k in extra_left})).splitlines()
         )
-    extra_right = set_right - set_left
+    extra_right = [k for k in right if k not in left]
     len_extra_right = len(extra_right)
     if len_extra_right:
         explanation.append(
**diff --git a/testing/test_assertion.py b/testing/test_assertion.py**
index 5179b13b0..a9cb1075d 100644
--- a/testing/test_assertion.py
+++ b/testing/test_assertion.py
@@ -2198,3 +2198,36 @@ def test_full_output_vvv(pytester: Pytester) -> None:
         ]
     )
     result.stdout.no_fnmatch_line(expected_non_vvv_arg_line)
+
+import re
+
+def test_dict_extra_items_preserve_insertion_order(pytester):
+    pytester.makepyfile(
+        test_order="""
+        def test_order():
+            a = {
+                "first": 1,
+                "second": 2,
+            }
+            assert a == {}
+        """
+    )
+
+    result = pytester.runpytest()
+
+    stdout = result.stdout.str()
+    stdout = re.sub(r"\x1b\[[0-9;]*m", "", stdout)
+
+    assert "Left contains 2 more items:" in stdout
+
+    dict_line = next(
+        line for line in stdout.splitlines()
+        if "{" in line and "}" in line and "first" in line
+    )
+
+    first = dict_line.find("first")
+    second = dict_line.find("second")
+
+    assert first != -1 and second != -1
+    assert first < second
+

RafiB982 commented:
Hi @nicoddemus, let me know if you are okay with PR..
